### ISSUE TITLE
access violation in bpf_object__open

### ISSUE THREAD

**Issue by mtfriesen** [microsoft/ebpf-for-windows#2131]
### Describe the bug

The XDP stress test crashed due to an access violation within `bpf_object__open`.

### OS information

`17763.1.amd64fre.rs5_release.180914-1434`

### Steps taken to reproduce bug

Run XDP stress test with eBPF integration.

### Expected behavior

The stress test should not crash.

### Actual outcome

The stress test crashed.

### Additional details

Crash dump: [ebpf_av_open.zip](https://github.com/microsoft/ebpf-for-windows/files/10818818/ebpf_av_open.zip)

```
0:011> kn
 # Child-SP          RetAddr               Call Site
00 000000e2`80ffc468 00007ffc`d593f5b1     ntdll!ZwAlpcSendWaitReceivePort+0x14 [minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 1283] 
01 000000e2`80ffc470 00007ffc`d593e9a7     ntdll!SendMessageToWERService+0x1a5 [minkernel\ntdll\wer.c @ 491] 
02 000000e2`80ffc5d0 00007ffc`d54d8133     ntdll!ReportExceptionInternal+0xe3 [minkernel\ntdll\wer.c @ 627] 
03 000000e2`80ffd140 00007ffc`d54d7d2c     kernel32!WerpReportFaultInternal+0x3cf [onecore\windows\feedback\core\faultrep\lib\faultrep.cpp @ 896] 
04 000000e2`80ffd260 00007ffc`d252725e     kernel32!WerpReportFault+0xac [onecore\windows\feedback\core\faultrep\lib\faultrep.cpp @ 1313] 
05 000000e2`80ffd2a0 00007ffc`d5909240     KERNELBASE!UnhandledExceptionFilter+0x35e [minkernel\kernelbase\xcpt.c @ 685] 
06 (Inline Function) --------`--------     ntdll!RtlpThreadExceptionFilter+0x99 [minkernel\ntdll\rtlstrt.c @ 967] 
07 000000e2`80ffd3b0 00007ffc`d58efe56     ntdll!RtlUserThreadStart$filt$0+0xaa [minkernel\ntdll\rtlstrt.c @ 1164] 
08 000000e2`80ffd3f0 00007ffc`d590582f     ntdll!__C_specific_handler+0x96 [minkernel\crts\crtw32\misc\riscchandler.c @ 352] 
09 000000e2`80ffd460 00007ffc`d5864cef     ntdll!RtlpExecuteHandlerForException+0xf [minkernel\ntos\rtl\amd64\xcptmisc.asm @ 132] 
0a 000000e2`80ffd490 00007ffc`d590459e     ntdll!RtlDispatchException+0x40f [minkernel\ntos\rtl\amd64\exdsptch.c @ 628] 
0b 000000e2`80ffdbc0 00007ffc`b3528f3d     ntdll!KiUserExceptionDispatch+0x2e [minkernel\ntos\rtl\amd64\trampoln.asm @ 749] 
0c (Inline Function) --------`--------     EbpfApi!ELFIO::elfio::Sections::operator[]+0x36 [D:\ebpf-for-windows\external\ebpf-verifier\external\elfio\elfio\elfio.hpp @ 962] 
0d (Inline Function) --------`--------     EbpfApi!ELFIO::symbol_section_accessor_template<ELFIO::section const >::find_hash_section+0x4a [D:\ebpf-for-windows\external\ebpf-verifier\external\elfio\elfio\elfio_symbols.hpp @ 268] 
0e (Inline Function) --------`--------     EbpfApi!ELFIO::symbol_section_accessor_template<ELFIO::section const >::{ctor}+0x75 [D:\ebpf-for-windows\external\ebpf-verifier\external\elfio\elfio\elfio_symbols.hpp @ 37] 
0f 000000e2`80ffe360 00007ffc`b352f5e2     EbpfApi!_get_program_and_map_names+0x18d [D:\ebpf-for-windows\libs\api\Verifier.cpp @ 71] 
10 000000e2`80ffe6b0 00007ffc`b350f66c     EbpfApi!load_byte_code+0x1152 [D:\ebpf-for-windows\libs\api\Verifier.cpp @ 271] 
11 000000e2`80ffe980 00007ffc`b35180d6     EbpfApi!_initialize_ebpf_object_from_elf+0x13c [D:\ebpf-for-windows\libs\api\ebpf_api.cpp @ 1772] 
12 (Inline Function) --------`--------     EbpfApi!_initialize_ebpf_object_from_file+0x5c [D:\ebpf-for-windows\libs\api\ebpf_api.cpp @ 1817] 
13 000000e2`80ffea60 00007ffc`b3534cb7     EbpfApi!ebpf_object_open+0x1e6 [D:\ebpf-for-windows\libs\api\ebpf_api.cpp @ 2307] 
14 (Inline Function) --------`--------     EbpfApi!bpf_object__open_file+0x54 [D:\ebpf-for-windows\libs\api\libbpf_object.cpp @ 101] 
15 000000e2`80ffeb20 00007ff7`9cfe2f12     EbpfApi!bpf_object__open+0x67 [D:\ebpf-for-windows\libs\api\libbpf_object.cpp @ 89] 
16 000000e2`80ffeb90 00000000`00000000     spinxsk!AttachXdpEbpfProgram+0x282 [D:\xdp\test\spinxsk\spinxsk.c @ 438] 
```

**Comment by mtfriesen** [comment]
Adding data from a different repro (with ASAN) since #2134 is presumably a duplicate of this issue:

Attached is the ebpfapi.dll and matching PDB to resolve proper symbols: [ebpfapi_symbols.zip](https://github.com/microsoft/ebpf-for-windows/files/10828800/ebpfapi_symbols.zip)

```
==2436==ERROR: AddressSanitizer: access-violation on unknown address 0x000000000000 (pc 0x7ffffcb98f3d bp 0x00ae0b8feb40 sp 0x00ae0b8fea40 T119)
==2436==The signal is caused by a READ memory access.
==2436==Hint: address points to the zero page.
    #0 0x7ffffcb98f3c in libbpf_num_possible_cpus+0x662c (C:\ebpf\EbpfApi.dll+0x180098f3c)
    #1 0x7ffffcb9f5e1 in libbpf_num_possible_cpus+0xccd1 (C:\ebpf\EbpfApi.dll+0x18009f5e1)
    #2 0x7ffffcb7f66b  (C:\ebpf\EbpfApi.dll+0x18007f66b)
    #3 0x7ffffcb880d5 in ebpf_object_get+0xd95 (C:\ebpf\EbpfApi.dll+0x1800880d5)
    #4 0x7ffffcba4cb6 in bpf_object__open+0x66 (C:\ebpf\EbpfApi.dll+0x1800a4cb6)
    #5 0x7ff774b328f9 in AttachXdpEbpfProgram D:\xdp\test\spinxsk\spinxsk.c:440
    #6 0x7ff774b33dab in AttachXdpProgram D:\xdp\test\spinxsk\spinxsk.c:563
    #7 0x7ff774b3b5e9 in FuzzSocketMisc D:\xdp\test\spinxsk\spinxsk.c:1451
    #8 0x7ff774b3ea68 in XskFuzzerWorkerFn D:\xdp\test\spinxsk\spinxsk.c:1885
    #9 0x7ff774b6c519 in __asan::AsanThread::ThreadStart D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_thread.cpp:262
    #10 0x7ff80ebc7ad3 in BaseThreadInitThunk+0x13 (C:\Windows\System32\KERNEL32.DLL+0x180017ad3)
    #11 0x7ff81137a370 in RtlUserThreadStart+0x20 (C:\Windows\SYSTEM32\ntdll.dll+0x18005a370)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: access-violation (C:\ebpf\EbpfApi.dll+0x180098f3c) in libbpf_num_possible_cpus+0x662c
Thread T119 created by T4 here:
    #0 0x7ff774b6af0f in __asan_wrap_CreateThread D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_win.cpp:149
    #1 0x7ff774b40b34 in QueueWorkerFn D:\xdp\test\spinxsk\spinxsk.c:2053
    #2 0x7ff774b6c519 in __asan::AsanThread::ThreadStart D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_thread.cpp:262
    #3 0x7ff80ebc7ad3 in BaseThreadInitThunk+0x13 (C:\Windows\System32\KERNEL32.DLL+0x180017ad3)
    #4 0x7ff81137a370 in RtlUserThreadStart+0x20 (C:\Windows\SYSTEM32\ntdll.dll+0x18005a370)

Thread T4 created by T0 here:
    #0 0x7ff774b6af0f in __asan_wrap_CreateThread D:\a\_work\1\s\src\vctools\asan\llvm\compiler-rt\lib\asan\asan_win.cpp:149
    #1 0x7ff774b440de in main D:\xdp\test\spinxsk\spinxsk.c:2411
    #2 0x7ff774b9ae38 in invoke_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
    #3 0x7ff774b9ad8d in __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
    #4 0x7ff774b9ac4d in __scrt_common_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:330
    #5 0x7ff774b9aead in mainCRTStartup D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp:16
    #6 0x7ff80ebc7ad3 in BaseThreadInitThunk+0x13 (C:\Windows\System32\KERNEL32.DLL+0x180017ad3)
    #7 0x7ff81137a370 in RtlUserThreadStart+0x20 (C:\Windows\SYSTEM32\ntdll.dll+0x18005a370)

==2436==ABORTING
```

**Comment by dthaler** [comment]
This issue is due to not supporting an ELF file with no `.symtab` section.

**Comment by dthaler** [comment]
@mtfriesen can you send the .o file that was being loaded?

**Comment by dthaler** [comment]
> This issue is due to not supporting an ELF file with no `.symtab` section.

Further analysis: 
Verifier.cpp line 187 reads the elf file.
If the load fails, or no `.symtab` section then asm_files.cpp line 120 should throw and never call the stack shown in this bug.
Otherwise line 269 calls `_get_program_and_map_names()`, so there must have been a `.symtab` section.
But in the crash at this point sections 8-16 have no name so it cannot find the `.symtab` section.
Perhaps they were freed or corrupted somehow?

**Comment by dahavey** [comment]
@mtfriesen - no repro, could you please attach the bpf program, symbols and repro again? thanks.

**Comment by mtfriesen** [comment]
Sure, I will reopen with the requested artifacts upon the next repro.

**Comment by mtfriesen** [comment]
@dthaler here is a new crash dump with bpf programs and symbols.

[ebpf6.zip](https://github.com/microsoft/ebpf-for-windows/files/11078888/ebpf6.zip)

**Comment by mtfriesen** [comment]
After upgrading to a newer eBPF build, this crash has not occurred. I'm closing this for now and will reactivate if we do get a repro.

**Comment by mtfriesen** [comment]
We have a repro on a recent build of eBPF in our CI/CD: https://azurewatson.microsoft.com/dumps/7169378350

Faulting module: `EbpfApi.dll 0.0.0.1 2023-04-09 07:52:39Z (64326EC7)`

```
ucrtbase!_invoke_watson+0x18 [minkernel\crts\ucrt\src\appcrt\misc\invalid_parameter.cpp @ 224]
ucrtbase!_invalid_parameter+0x81 [minkernel\crts\ucrt\src\appcrt\misc\invalid_parameter.cpp @ 112]
ucrtbase!_invalid_parameter_noinfo+0x9 [minkernel\crts\ucrt\src\appcrt\misc\invalid_parameter.cpp @ 118]
ucrtbase!_close+0x3a381 [minkernel\crts\ucrt\src\appcrt\lowio\close.cpp @ 64]
ucrtbase!_fclose_nolock+0x47 [minkernel\crts\ucrt\src\appcrt\stdio\fclose.cpp @ 61]
ucrtbase!fclose+0x40 [minkernel\crts\ucrt\src\appcrt\stdio\fclose.cpp @ 36]
ebpfapi!std::basic_filebuf_char,std::char_traits_char_ _::close+0x69 [C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.34.31933\include\fstream @ 392]
ebpfapi!std::basic_filebuf_char,std::char_traits_char_ _::~basic_filebuf_char,std::char_traits_char_ _+0x5a [C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.34.31933\include\fstream @ 186]
ebpfapi!std::basic_ifstream_char,std::char_traits_char_ _::`vbase destructor'+0x45
ebpfapi!read_elf+0xc7
ebpfapi!load_byte_code+0x1d4 [D:\a\ebpf-for-windows\ebpf-for-windows\libs\api\Verifier.cpp @ 187]
ebpfapi!_initialize_ebpf_object_from_elf+0x13c [D:\a\ebpf-for-windows\ebpf-for-windows\libs\api\ebpf_api.cpp @ 1779]
ebpfapi!ebpf_object_open+0x1e6 [D:\a\ebpf-for-windows\ebpf-for-windows\libs\api\ebpf_api.cpp @ 2314]
ebpfapi!bpf_object__open+0x67 [D:\a\ebpf-for-windows\ebpf-for-windows\libs\api\libbpf_object.cpp @ 91]
```[EbpfApi.dll](https://azurewatson.microsoft.com/?FaultingModule=EbpfApi.dll) [0.0.0.1](https://azurewatson.microsoft.com/?FaultingModuleVer=0.0.0.1) 2023-04-09 07:52:39Z (64326EC7)

**Comment by saxena-anurag** [comment]
It looks like the above one is a different call stack than the original issue. I found another instance of the original issue -- https://azurewatson.microsoft.com/dumps/7191372087 where it again there is no `.symtab` section. But not sure how that is possible, as it should have failed earlier as Dave mentioned earlier. 

Maybe this is a bug in ELFIO when we are loading same .o file in parallel?

**Comment by mtfriesen** [comment]
> @mtfriesen Is it possible to get a local repro (and a debugger) for this issue?

It will be hard to repro this on demand, at least right now, because our stress tests are hitting a hang or deadlock in eBPF (#2457) much more frequently than anything else.

In principle, the repro steps are fairly straightforward: [run the XDP spinxsk test](https://github.com/microsoft/xdp-for-windows/blob/main/docs/development.md#running-spinxsk) with eBPF enabled. Here's a command that produces minimal noise, works in the most environments, and enables eBPF: `.\tools\spinxsk.ps1 -SuccessThresholdPercent 1 -XdpmpPollProvider FNDIS -EnableEbpf`

**Comment by saxena-anurag** [comment]
@mtfriesen is this issue still seen the XDP CI tests?

**Comment by mtfriesen** [comment]
Nope. I'm on a 0.9.0 build and haven't root caused any bugs to eBPF. Closing as not repro.
